tests/h/services/user_signup_test.py

```diff
@@ -130,6 +131,39 @@ def test_signup_records_stats_if_present(self, svc, stats):
 
         stats.incr.assert_called_once_with("auth.local.register")
 
+    def test_signup_logs_conflict_error_when_account_with_email_already_exists(
+        self, svc, db_session, patch
+    ):
+        log = patch("h.services.user_signup.log")
+
+        with pytest.raises(ConflictError):
+            svc.signup(username="foo", email="foo@bar.com")
+            svc.signup(username="foo", email="foo@bar.com")
+
+        log.warning.assert_called_once_with(
+            "email conflict error occured during user signup"
+        )
+
+    def test_signup_raises_conflict_error_when_account_with_email_already_exists(
+        self, svc, db_session
+    ):
+        with pytest.raises(
+            ConflictError,
+            match="The email account foo@bar.com has already been registered.",
+        ):
+            svc.signup(username="foo", email="foo@bar.com")
+            svc.signup(username="foo", email="foo@bar.com")
+
+    def test_signup_does_not_raise_conflict_error_when_non_email_related_integrity_error_occurs(
+        self, svc, db_session
+    ):
+        with pytest.raises(
+            IntegrityError,
+            match='duplicate key value violates unique constraint "ix__user__userid"',
+        ):
```

Comment:
Hmm ... if a user submits the same registration multiple times it seems likely that both the username AND the email would conflict. In that case, do we know which error Postgres is going to detect first? Could that change in future?

Reply:
The email will be detected first which is why we're catching on email here. The constraint is defined in the table definition so in order for this to change, we would have to change the table definition. Postgres evaluates the (authority, email) combined unique constraint index first. I don't anticipate that that will change in the future.

Reply:
> The email will be detected first which is why we're catching on email here.

Why is that and how do you know it couldn't change in future?

Reply:
The order in which it performs the checks happens within postgres. There is some documentation on it here: https://www.postgresql.org/docs/9.4/index-unique-checks.html. As far as I can tell the uniqueness checks haven't really changed from 9.4-11 so I think it's unlikely they will change in the future. I haven't been able to track down any official documentation on the order in which checks happen although it is consistently the case here that the combinational unique index check happens before the individual unique column checks.

Reply:
> As far as I can tell the uniqueness checks haven't really changed from 9.4-11 so I think it's unlikely they will change in the future.

I think you are probably right here, but I still feel uncomfortable about relying on implicit behaviour that we can't be sure is guaranteed. If you can find a way to tolerate changes in behaviour (eg. in case the username conflict was reported instead) without complicating the code much (perhaps even simplifying it) I think that would be a good idea.

Reply:
(Just a thought from the Peanut Gallery here). Our code is nominally abstracted from the DB of choice (postgres). So it would make sense for our code not to be implicitly dependent on postgres-specific implementation details.